packages/@lwc/engine/src/framework/template.ts

```diff
@@ -132,6 +133,10 @@ export function evaluateTemplate(vm: VM, html: Template): Array<VNode | null> {
                 html
             )}`
         );
+        // in dev-mode, we support hot swapping of templates, which means that
+        // the component instance might be attempting to use an old version of
+        // the template, while internally, we have a replacement for it.
+        html = getTemplateOrSwappedTemplate(html);
```

Comment:
this returns the same template provided, or the new version of it based on calls to `registerTemplateSwap`